### PR TITLE
Json operations - Part I

### PR DESCRIPTION
**Pull request by SamMarkowitz**

### CODE REVIEW

File: content/io/cloudslang/base/json/add_value.sl [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+# Adds a value to the given JSON at the key represented by the key_list if one doesn't already exist
+# or replaces the value at the key represented by the key_list if one already exists
+#
+# Inputs:
+#   - json_input - JSON data input
+#   - key_list - list of keys to add - Example: ['tags', 1, 'name']
```
Comment (oritstone) [inline content/io/cloudslang/base/json/add_value.sl]: Need to document better that the user should give a Json path to the object. Also the input name should probably be different + the operation description above

File: content/io/cloudslang/base/json/add_value.sl [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        else:
+          decoded = value
+        encoded = json.dumps(decoded)
+        return_code = '0'
+        return_result = 'Parsing successful.'
+      except Exception as ex:
+        return_result = ex
+        return_code = '-1'
+  outputs:
+    - json_output: encoded if return_code == '0' else ''
```
Comment (oritstone) [inline content/io/cloudslang/base/json/add_value.sl]: you can use our new get() function. It is more elegant

File: content/io/cloudslang/base/json/get_keys.sl [hunk 3]
```diff
@@ -0,0 +1,50 @@
+#   (c) Copyright 2015 Hewlett-Packard Development Company, L.P.
+#   All rights reserved. This program and the accompanying materials
+#   are made available under the terms of the Apache License v2.0 which accompany this distribution.
+#
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+# Parses the given JSON input to retrieve the corresponding value addressed by the keys_list input.
```
Comment (oritstone) [inline content/io/cloudslang/base/json/get_keys.sl]: Seems to me that the description of the op + all the inputs outputs are wrong...

File: test/io/cloudslang/base/json/add_value.inputs.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+testAddValue:
+  inputs:
+    - json_before: '{"outer_key1": "outer_value1", "outer_key2": {"inner_key1": "inner_value1", "inner_key2": ["list_item1", "list_item2"]}}'
+    - key_list: [outer_key2, inner_key3]
+    - value: inner_value3
+    - json_after: '{"outer_key2": {"inner_key1": "inner_value1", "inner_key3": "inner_value3", "inner_key2": ["list_item1", "list_item2"]}, "outer_key1": "outer_value1"}'
+  description: Tests that add_value.sl correctly adds value to JSON
+  testFlowPath: io.cloudslang.base.json.test_add_value
+  result: SUCCESS
```
Comment (oritstone) [inline test/io/cloudslang/base/json/add_value.inputs.yaml]: Maybe you should add test case to add a value to an existing key path, not creating a new key

File: test/io/cloudslang/base/json/get_value.inputs.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    - value: inner_value1
+
+testGetValueInnerArray:
+  inputs:
+    - json_input: '{"outer_key1": "outer_value1", "outer_key2": {"inner_key1": "inner_value1", "inner_key2": ["list_item1", "list_item2"]}}'
+    - key_list: [outer_key2, inner_key2, 1]
+  description: Tests that get_value.sl correctly gets an inner array value from a JSON
+  testFlowPath: io.cloudslang.base.json.get_value
+  outputs:
+    - value: list_item2
```
Comment (oritstone) [inline test/io/cloudslang/base/json/get_value.inputs.yaml]: Add a test case that returns as the value the entire array

File: test/io/cloudslang/base/json/test_add_value.sl [hunk 6]
```diff
@@ -0,0 +1,47 @@
+#   (c) Copyright 2015 Hewlett-Packard Development Company, L.P.
```
Comment (oritstone) [inline test/io/cloudslang/base/json/test_add_value.sl]: This flow should be added as a test flow only, and not to our content. Should be added under: `test/io/cloudslang/testbase/json`

**Comment by SamMarkowitz** [comment]
Still working on a couple of tests that are failing because of some character encoding issues (I think).

File: content/io/cloudslang/base/json/add_value.sl [hunk 7]
```diff
@@ -0,0 +1,59 @@
+#   (c) Copyright 2015 Hewlett-Packard Development Company, L.P.
+#   All rights reserved. This program and the accompanying materials
+#   are made available under the terms of the Apache License v2.0 which accompany this distribution.
+#
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+# Adds or replaces a value to the given JSON at the key or index represented by the json_path.
```
Comment (oritstone) [inline content/io/cloudslang/base/json/add_value.sl]: Adds or replaces a value to the given JSON at the `keys or indices` represented by the json_path

File: content/io/cloudslang/base/json/get_keys.sl [hunk 8]
```diff
@@ -0,0 +1,52 @@
+#   (c) Copyright 2015 Hewlett-Packard Development Company, L.P.
+#   All rights reserved. This program and the accompanying materials
+#   are made available under the terms of the Apache License v2.0 which accompany this distribution.
+#
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+# Retrieves keys from a JSON object.
```
Comment (oritstone) [inline content/io/cloudslang/base/json/get_keys.sl]: Maybe add something like: Retrieves keys from a JSON object `at a given JSON path`

File: content/io/cloudslang/base/json/add_value.sl [hunk 9]
```diff
@@ -0,0 +1,59 @@
+#   (c) Copyright 2015 Hewlett-Packard Development Company, L.P.
+#   All rights reserved. This program and the accompanying materials
+#   are made available under the terms of the Apache License v2.0 which accompany this distribution.
+#
+#   The Apache License is available at
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+####################################################
+# Adds or replaces a value to the given JSON at the keys or indices represented by the json_path.
+# If the full path does not exist, the keys for the path are added as well.
```
Comment (Bonczidai) [inline content/io/cloudslang/base/json/add_value.sl]: only the last key is added to the path, right? e.g. ['exist', 'dont_exist_1', 'dont_exist_2'] will fail

Comment (SamMarkowitz) [inline content/io/cloudslang/base/json/add_value.sl]: Good catch!

What's your opinion:

1. Stay with this functionality and document it
2. Don't allow adding for any non-existent keys
3. Add all keys in path that don't exist

Comment (Bonczidai) [inline content/io/cloudslang/base/json/add_value.sl]: I prefer the first one: updating the doc

Comment (SamMarkowitz) [inline content/io/cloudslang/base/json/add_value.sl]: Updated

Comment (Bonczidai) [inline content/io/cloudslang/base/json/add_value.sl]: :+1:

File: content/io/cloudslang/base/json/add_value.sl [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+#   - json_path - path at which to add value represented as a list of keys and/or indices - Example: ['tags', 1, 'name']
+#   - value - value to associate with key
+# Outputs:
+#   - json_output - JSON with keys:value added
+#   - return_result - parsing was successful or not
+#   - return_code - "0" if parsing was successful, "-1" otherwise
+#   - error_message - error message if there was an error when executing, empty otherwise
+# Results:
+#   - SUCCESS - parsing was successful (return_code == '0') and value was added
+#   - FAILURE - parsing was unsucceful or the path does not exist
```
Comment (mihaitusa) [inline content/io/cloudslang/base/json/add_value.sl]: typo in "unsucceful"

**Comment by mihaitusa** [comment]
@SamMarkowitz  Everything looks ok. I think you can merge it. Nice work :+1:

**Pull request merged**